crates/sui-framework/src/natives/crypto/hmac.rs

```diff
@@ -13,20 +14,57 @@ use move_vm_types::{
 use smallvec::smallvec;
 use std::collections::VecDeque;
 
+const HMAC_SHA3_256_BLOCK_SIZE: usize = 136;
+
+#[derive(Clone)]
+pub struct HmacHmacSha3256CostParams {
+    /// Base cost for invoking the `hmac_sha3_256` function
+    pub hmac_hmac_sha3_256_cost_base: InternalGas,
+    ///  Cost per byte of `msg` and `key`
+    pub hmac_hmac_sha3_256_input_cost_per_byte: InternalGas,
+    ///  Cost per block of `msg` and `key`, with block size = 136
+    pub hmac_hmac_sha3_256_input_cost_per_block: InternalGas,
+}
 pub fn hmac_sha3_256(
-    _context: &mut NativeContext,
+    context: &mut NativeContext,
     ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
     debug_assert!(ty_args.is_empty());
     debug_assert!(args.len() == 2);
 
+    // Load the cost paramaters from the protocol config
+    let hmac_hmac_sha3_256_cost_params = &context
+        .extensions()
+        .get::<NativesCostTable>()
+        .hmac_hmac_sha3_256_cost_params
+        .clone();
+
+    // Charge the base cost for this oper
+    native_charge_gas_early_exit!(
+        context,
+        hmac_hmac_sha3_256_cost_params.hmac_hmac_sha3_256_cost_base
+    );
+
     let message = pop_arg!(args, VectorRef);
     let key = pop_arg!(args, VectorRef);
-    let hmac_key = hmac::HmacKey::from_bytes(&key.as_bytes_ref()).unwrap();
 
-    // TODO: implement native gas cost estimation https://github.com/MystenLabs/sui/issues/3593
-    let cost = legacy_empty_cost();
+    let msg_len = message.as_bytes_ref().len();
+    let key_len = key.as_bytes_ref().len();
+    // Charge the arg size dependent costs
+    native_charge_gas_early_exit!(
+        context,
+        hmac_hmac_sha3_256_cost_params.hmac_hmac_sha3_256_input_cost_per_byte
+            // same cost for msg and key
+            * ((msg_len + key_len) as u64).into()
+            + hmac_hmac_sha3_256_cost_params.hmac_hmac_sha3_256_input_cost_per_block
+                * ((((msg_len + key_len) + (2 * HMAC_SHA3_256_BLOCK_SIZE - 2))
+                    / HMAC_SHA3_256_BLOCK_SIZE) as u64)
```

Comment:
each of those should be rounded independently to HMAC_SHA3_256_BLOCK_SIZE

Reply:
Yes thats the idea behind the `(2 * HMAC_SHA3_256_BLOCK_SIZE - 2)`. It has the contribution of both components.